resources/ext/cli/config-migration

```diff
@@ -0,0 +1,243 @@
+#!/bin/bash
+
+migration_command="puppetdb config-migration"
+
+#############
+# FUNCTIONS #
+#############
+
+# Backs up a file, if it hasn't already been backed up
+#
+# $1 - file to backup
+#
+# Example:
+#
+#   backupfile "/etc/myconfig"
+#
+function backupfile() {
+  # Create the global array if it doesn't already exist
+  if [ -z "$backupfile_list" ]; then
+    backupfile_list=()
+  fi
+
+  # We check the array to make sure the file isn't already backed up
+  if ! contains ${backupfile_list[@]} "$1"; then
+    local backup_path="$1.bak.$(date +%s)"
+    echo "Backing up $1 to ${backup_path} before making changes"
+    cp -p "$1" "$backup_path"
```

Comment:
I'd recommend checking the exit status here (imagine disk full, etc.), and more generally, I'd suggest checking the exit status of every relevant shell command, whether via explicit "if ! cp -p ..." or "set -e".

Though beware "set -e"'s precise behavior, i.e. this command will actually print 13: `(set -e; false; echo 13) || false` and IIRC the same is true for `foo || false` if foo is a function that tries to use set -e in a subshell.  The details are explained here: https://www.gnu.org/software/bash/manual/bashref.html#The-Set-Builtin

I'd also suggest considering use of "set -o pipefail" (with or without "set -e"), but we'd then need to be careful about situations where a non-zero exit status in a pipeline shouldn't be considered an error (i.e. grep in something like `foo | grep 'might or might not be found' | wc -l`).

Oh, and you may well already know this, but another thing that's bitten me in the past is the behavior of statements like `local foo=$(bar)` because at the time I hadn't realized that the `local` exit status is completely independent of the subcommand's exit status, i.e. after `local foo=$(false)`, the exit status variable $? will be 0.  I believe the same may be true for the other declarations as well.  To avoid that I'll often just handle the declarations separately, i..e. `local foo bar; foo=...`.

Finally, "set -u" can be interesting, but of course you have to be willing to explicitly initialize everything in the affected regions.

Hope this helps.